### PR TITLE
Automatically deploy JS legacy code coverage

### PR DESCRIPTION
**Pull request by tamarasaurus**
**Description (for Contributor and Core Developer)**

This PR is the real implementation of the previous POC to collect and deploy JS code coverage reports during behat tests - https://github.com/akeneo/pim-community-dev/pull/10605.

### How does it work?

Steps to capture the coverage:
1. Build webpack with coverage listeners injected into each module using `istanbul-instrument-loader`
2. Run all the behats
3. Fetch the instrumentation report from the window variable after each scenario/page reload and write to file
4. Merge the coverage reports at the end and generate the coverage html
5. Compress the artifacts and make them available after each nightly
6. Push the HTML report to http://coverage.core.akeneo.com/

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | OK
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: tests/legacy/features/Behat/Context/HookContext.php [hunk 1]
```diff
@@ -262,6 +262,28 @@ protected function getPimFilesystems()
...
+        if ($event->getTestResult() !== StepResult::UNDEFINED || $event === null) {
+            if ($this->getSession()->getDriver() instanceof Selenium2Driver) {
+                try {
+                    $script = 'return window.coverage;';
+                    $coverage = $this->getSession()->evaluateScript($script);
+                    $path = sprintf('%s/%s.json', './.coverage', uniqid());
+                    $fs = new \Symfony\Component\Filesystem\Filesystem();
+                    $fs->dumpFile($path, json_encode($coverage));
+                } catch (\Exception $e) {
+                    var_dump($e->getMessage());
```
Comment (ahocquard) [inline tests/legacy/features/Behat/Context/HookContext.php]: I think we should not catch the error to be properly logged

File: .circleci/config.yml [hunk 2]
```diff
@@ -210,6 +210,39 @@ jobs:
...
+
+  collect_js_coverage:
+    machine:
+      image: ubuntu-1604:201903-01
+    steps:
+      - attach_workspace:
+          at: ~/
+      - add_ssh_keys:
+          fingerprints:
+            - "3b:a2:61:e2:63:33:69:5e:29:9b:c4:92:e9:77:71:c5"
```
Comment (mmetayer) [inline .circleci/config.yml]: shouldn't we put these in a CIRCLECI environment variable?

File: .circleci/config.yml [hunk 3]
```diff
@@ -210,6 +210,39 @@ jobs:
...
+          name: Merge JS code coverage reports and output HTML
+          command: |
+              docker-compose run --rm node yarn run instrument-merge
+              docker-compose run --rm node yarn run instrument-report
+      - run:
+          name: Compress HTML coverage report
+          command: tar -cvf coverage.tar ./coverage/
+      - run:
+          name: Deploy coverage report
+          command: scp -P 2202 -r ./coverage/* akeneo@coverage.core.akeneo.com:/var/www/html/frontend-ce/
```
Comment (mmetayer) [inline .circleci/config.yml]: same comment

Comment (ahocquard) [reply .circleci/config.yml]: we have to check if it's secure (if we give access to other people to our CI for contribution, a simple echo could compromise this kind of obfuscation)

**Pull request merged**